File: solutions/zig/linked-list/1/linked_list.zig
pub fn LinkedList(comptime T: type) type {
    return struct {
        const Self = @This();

        pub const Node = struct {
            prev: ?*Node = null,
            next: ?*Node = null,
            data: T,
        };

        first: ?*Node = null,
        last: ?*Node = null,
        len: usize = 0,

        pub fn push(self: *Self, node: *Node) void {
            if (self.last) |tail| {
                node.prev = tail;
                tail.next = node;
            } else {
                self.first = node;
            }
            self.last = node;
            self.len += 1;
        }

        pub fn pop(self: *Self) ?*Node {
            if (self.last) |tail| {
                self.last = tail.prev;
                tail.prev = null;
                if (self.last) |_tail| {
                    _tail.next = null;
                } else {
                    self.first = null;
                }
                self.len -= 1;
                return tail;
            } else {
                return null;
            }
        }

        pub fn shift(self: *Self) ?*Node {
            if (self.first) |head| {
                self.first = head.next;
                head.next = null;
                if (self.first) |_head| {
                    _head.prev = null;
                } else {
                    self.last = null;
                }
                self.len -= 1;
                return head;
            } else {
                return null;
            }
        }

        pub fn unshift(self: *Self, node: *Node) void {
            if (self.first) |head| {
                head.prev = node;
                node.next = head;
            } else {
                self.last = node;
            }
            self.first = node;
            self.len += 1;
        }

        pub fn delete(self: *Self, node: *Node) void {
            var iter = self.first;
            while (iter) |curr| : (iter = curr.next) {
                if (curr == node) {
                    break;
                }
            }
            var _node = iter orelse return;
            if (_node.next) |next| {
                if (_node.prev) |prev| {
                    next.prev = prev;
                    prev.next = next;
                    _node.next = null;
                    _node.prev = null;
                } else {
                    self.first = _node.next;
                    if (self.last == node) {
                        self.last = null;
                    }
                }
            } else {
                self.last = _node.prev;
                if (self.first == node) {
                    self.first = null;
                }
            }
            self.len -= 1;
        }
    };
}

File: solutions/go/linked-list/1/linked_list.go
package linkedlist

import "errors"

var errEmptyList error = errors.New("Empty List")

type Node struct {
	Value    any
	next     *Node
	previous *Node
}

func (n *Node) Next() *Node {
	return n.next
}

func (n *Node) Prev() *Node {
	return n.previous
}

type List struct {
	head *Node
	tail *Node
}

func NewList(elements ...any) *List {
	l := &List{}
	for _, el := range elements {
		l.Push(el)
	}
	return l
}

func (l *List) Unshift(v any) {
	n := &Node{Value: v}
	if l.head == nil {
		l.head, l.tail = n, n
		return
	}
	n.next, l.head.previous, l.head = l.head, n, n
}

func (l *List) Push(v any) {
	n := &Node{Value: v}
	if l.tail == nil {
		l.head, l.tail = n, n
		return
	}
	n.previous, l.tail.next, l.tail = l.tail, n, n
}

func (l *List) Shift() (any, error) {
	if l.head == nil {
		return nil, errEmptyList
	}
	n := l.head

	if l.head == l.tail {
		l.head, l.tail = nil, nil
	} else {
		l.head.next.previous, l.head = nil, n.next
	}
	return n.Value, nil
}

func (l *List) Pop() (any, error) {
	if l.head == nil {
		return nil, errEmptyList
	}
	n := l.tail
	if l.tail == l.head {
		l.head, l.tail = nil, nil
	} else {
		l.tail.previous.next, l.tail = nil, n.previous
	}
	return n.Value, nil
}

func (l *List) Reverse() {
	l.head, l.tail = l.tail, l.head
	for iter := l.head; iter != nil; iter = iter.next {
		iter.next, iter.previous = iter.previous, iter.next
	}
}

func (l *List) First() *Node {
	return l.head
}

func (l *List) Last() *Node {
	return l.tail
}
